server/lib/src/idm/application.rs

```diff
@@ -0,0 +1,715 @@
+        })
+    }
+
+    pub fn commit(self) {
+        self.inner.commit();
+    }
+}
+
+impl LdapApplications {
+    pub fn read(&self) -> LdapApplicationsReadTransaction {
+        LdapApplicationsReadTransaction {
+            inner: self.inner.read(),
+        }
+    }
+
+    pub fn write(&self) -> LdapApplicationsWriteTransaction {
+        LdapApplicationsWriteTransaction {
+            inner: self.inner.write(),
+        }
+    }
+}
+
+impl TryFrom<Vec<Arc<EntrySealedCommitted>>> for LdapApplications {
+    type Error = OperationError;
+
+    fn try_from(value: Vec<Arc<EntrySealedCommitted>>) -> Result<Self, Self::Error> {
+        let apps = LdapApplications {
+            inner: CowCell::new(LdapApplicationsInner {
+                set: HashMap::new(),
+            }),
+        };
+
+        let mut apps_wr = apps.write();
+        apps_wr.reload(value)?;
+        apps_wr.commit();
+        Ok(apps)
+    }
+}
+
+impl<'a> IdmServerAuthTransaction<'a> {
+    pub async fn application_auth_ldap(
+        &mut self,
+        lae: &LdapApplicationAuthEvent,
+        ct: Duration,
+    ) -> Result<Option<LdapBoundToken>, OperationError> {
+        let usr_entry = self.get_qs_txn().internal_search_uuid(lae.target)?;
+
+        let within_valid_window = Account::check_within_valid_time(
+            ct,
+            usr_entry
+                .get_ava_single_datetime(Attribute::AccountValidFrom)
+                .as_ref(),
+            usr_entry
+                .get_ava_single_datetime(Attribute::AccountExpire)
+                .as_ref(),
+        );
+
+        if !within_valid_window {
+            security_info!("Account has expired or is not yet valid, not allowing to proceed");
+            return Err(OperationError::SessionExpired);
+        }
+
+        let account: Account =
+            Account::try_from_entry_ro(&usr_entry, &mut self.qs_read).map_err(|e| {
+                admin_error!("Failed to search account {:?}", e);
+                e
+            })?;
+
+        if account.is_anonymous() {
+            return Err(OperationError::InvalidUuid);
+        }
+
+        let application = self
+            .ldap_applications
+            .inner
+            .set
+            .get(&lae.application)
+            .ok_or_else(|| {
+                admin_info!("Application {:?} not found", lae.application);
+                OperationError::NoMatchingEntries
+            })?;
+
+        // Check linked group membership
+        let linked_group_uuid = match application.linked_group {
+            Some(u) => u,
+            None => {
+                admin_warn!(
+                    "Application {:?} does not have a linked group.",
+                    application.name
+                );
+                return Ok(None);
+            }
+        };
+
+        let mut is_memberof = false;
+        if let Some(mut riter) = usr_entry.get_ava_as_refuuid(Attribute::MemberOf) {
+            if riter.any(|g| g == linked_group_uuid) {
```

Comment:
There are very few times where you should be using `any(|| )`. Generally it's a sign that you have a the incorrect type. You should be using https://github.com/kanidm/kanidm/blob/master/server/lib/src/entry.rs#L2556 because then you can do:

```
let is_memberof = usr_entry.get_ava_refer(Attribute::MemberOf)
    .map(|member_of_set| member_of_set.contains(linked_group_uuid))
    .unwrap_or_default();
```

Reply:
Fixed, thanks for the tip.